src/vcon/MarketGovernance.sol

```diff
@@ -0,0 +1,653 @@
+
+        unchecked {
+            for (uint256 i = 0; i < totalVenues; i++) {
+                address venue = pcvDeposits[i];
+                deposits[i].deposit = venue;
+                deposits[i].amount = getExpectedVenuePCVAmount(
+                    venue,
+                    totalPcv,
+                    cachedVconStaked
+                );
+            }
+        }
+    }
+
+    function getExpectedVenuePCVAmount(
+        address venue,
+        uint256 totalPcv,
+        uint256 totalVconStaked
+    ) public view returns (uint256 expectedPcvAmount) {
+        uint256 venueDepositedVcon = sharesToVcon(
+            venue,
+            venueTotalShares[venue]
+        );
+
+        if (totalVconStaked == 0) {
+            return 0;
+        }
+
+        expectedPcvAmount = (venueDepositedVcon * totalPcv) / totalVconStaked;
+    }
+
+    /// ------------- Helper Methods -------------
+
+    /// @param source address to pull funds from
+    /// @param destination recipient address for funds
+    /// @param swapper address to swap tokens with
+    /// @param amountPcv the amount of PCV to move from source
+    function _movePCVWithChecks(
+        address source,
+        address destination,
+        address swapper,
+        uint256 amountPcv,
+        uint256 totalPcv,
+        int256 sourceVenueBalance,
+        int256 destinationVenueBalance,
+        uint256 totalVconStaked
+    ) private {
+        address sourceAsset = IPCVDepositV2(source).balanceReportedIn();
+        address destinationAsset = IPCVDepositV2(destination)
+            .balanceReportedIn();
+
+        /// validate pcv movement
+        /// check underlying assets match up and if not that swapper is provided and valid
+        PCVRouter(pcvRouter).movePCV(
+            source,
+            destination,
+            swapper,
+            amountPcv,
+            sourceAsset,
+            destinationAsset
+        );
+
+        /// if nothing is staked on source, ignore balance check
+        if (sharesToVcon(source, venueTotalShares[source]) > 0) {
+            int256 sourceVenueBalanceAfter = getVenueDeviation(
+                source,
+                totalPcv,
+                totalVconStaked
+            );
+
+            /// source and dest venue balance measures the distance from being perfectly balanced
+
+            /// validate source venue balance became more balanced
+            _checkBalance(
+                sourceVenueBalance,
+                sourceVenueBalanceAfter,
+                "MarketGovernance: src more imbalanced"
+            );
+        }
+
+        int256 destinationVenueBalanceAfter = getVenueDeviation(
+            destination,
+            totalPcv,
+            totalVconStaked
+        );
+
+        /// validate destination venue balance became more balanced
+        _checkBalance(
+            destinationVenueBalance,
+            destinationVenueBalanceAfter,
+            "MarketGovernance: dest more imbalanced"
+        );
```

Comment:
I think you can simplify these checks by just doing absolute USD checks, no need for percentages of deviation. You can compare the "allocation error" (balance of the venue before & after vs the ideal balance), and require that it decreases. Then you're doing only simple operations & no multiplications & no need to work with ratios : 
```
uint256 idealSourceVenueBalance = getExpectedVenuePCVAmount(...);
uint256 sourceVenueBalanceBefore = PCVDeposit.balance();
uint256 sourceVenueBalanceAfter = PCVDeposit.balance();
uint256 allocationErrorBefore = sourceVenueBalanceBefore > idealSourceVenueBalance  ? sourceVenueBalanceBefore - idealSourceVenueBalance  : idealSourceVenueBalance - sourceVenueBalanceBefore;
uint256 allocationErrorAfter = sourceVenueBalanceAfter > idealSourceVenueBalance  ? sourceVenueBalanceAfter - idealSourceVenueBalance  : idealSourceVenueBalance - sourceVenueBalanceAfter;
require(allocationErrorAfter < allocationErrorBefore, "source venue became more imbalanced")
```

Reply:
Currently these checks are doing absolute USD checks, not looking at percentage deviation differences. See updated function `getVenueDeviation`, it returns the positive amount a venue is overweight by, or a negative value the venue is underweight by in raw PCV terms.

Reply:
ah yes ok perfect